src/main.cpp

```diff
@@ -23,35 +23,54 @@
 #include <QString>
 #include <QTextCodec>
 
-#include "mixxx.h"
+#include "coreservices.h"
 #include "mixxxapplication.h"
 #include "sources/soundsourceproxy.h"
 #include "errordialoghandler.h"
 #include "util/cmdlineargs.h"
 #include "util/console.h"
 #include "util/logging.h"
 #include "util/version.h"
+#include "widget/wmainwindow.h"
 
 #ifdef Q_OS_LINUX
 #include <X11/Xlib.h>
 #endif
 
 namespace {
 
+#define CLEAR_AND_CHECK_DELETED(x) clearHelper(x, #x);
+
+template <typename T>
+void clearHelper(std::shared_ptr<T>& ref_ptr, const char* name) {
+    std::weak_ptr<T> weak(ref_ptr);
+    ref_ptr.reset();
+    if (auto shared = weak.lock()) {
+        qWarning() << name << "was leaked!";
+    }
+}
+
 int runMixxx(MixxxApplication* app, const CmdlineArgs& args) {
     int result = -1;
-    MixxxMainWindow mainWindow(app, args);
-    // If startup produced a fatal error, then don't even start the
-    // Qt event loop.
-    if (ErrorDialogHandler::instance()->checkError()) {
-        mainWindow.finalize();
-    } else {
-        qDebug() << "Displaying main window";
-        mainWindow.show();
-
-        qDebug() << "Running Mixxx";
-        result = app->exec();
+    mixxx::CoreServices::initializeStaticServices();
+    {  // Scope for mixxx::CoreServices.
+        auto core = std::make_shared<mixxx::CoreServices>(nullptr, app, args.getSettingsPath());
+        core->initialize();
+        {  // Scope for WMainWindow.
+            WMainWindow window(app, core);
```

Comment:
This introduces a regression. No GUI is shown for about 3 seconds, then when the main window is first shown, the loading progress bar is already over half way.

Reply:
Ping. I propose moving the construction of services in the CoreServices class from the constructor to a new method that would get called by WMainWindow. Then have CoreServices emit a signal after each service is started that triggers a progress bar update in WMainWindow.